pil-analyzer/src/pil_analyzer.rs

```diff
@@ -317,9 +322,54 @@ impl PILAnalyzer {
         }
     }
 
+    pub fn traits_resolution(&mut self) {
+        fn collect_references(expr: &mut Expression) -> Vec<&mut Reference> {
+            match expr {
+                Expression::Reference(_, ref mut r @ Reference::Poly(_)) => vec![r],
+                _ => expr.children_mut().flat_map(collect_references).collect(),
+            }
+        }
+
+        let traits_functions_defs: HashMap<String, FunctionValueDefinition> = self
+            .definitions
+            .iter()
+            .filter_map(|(name, (_, def))| match def {
+                Some(FunctionValueDefinition::TraitFunction(_, _)) => {
+                    Some((name.clone(), def.clone().unwrap()))
+                }
+                _ => None,
+            })
+            .collect();
+
+        let mut references: Vec<&mut Reference> = self
+            .definitions
+            .iter_mut()
+            .flat_map(|(_, (_, def))| match def {
+                Some(FunctionValueDefinition::Expression(TypedExpression { e: expr, .. })) => {
+                    collect_references(expr)
+                }
+                _ => Vec::new(),
+            })
+            .chain(self.identities.iter_mut().flat_map(|identity| {
+                identity
+                    .left
+                    .selector
+                    .iter_mut()
+                    .chain(once(identity.left.expressions.as_mut()))
+                    .chain(identity.right.selector.iter_mut())
+                    .chain(once(identity.right.expressions.as_mut()))
+                    .flat_map(collect_references)
+            }))
+            .collect();
+
+        TraitsProcessor::new(&self.implementations)
```

Comment:
If we don't do anything with the TraitsProcessor object, maybe it would be better to define a public function in the traits_processor.rs and call that here without creating an object (and making the TraitsProcessor private)

Reply:
Also if you create the iterator here, why not call trait resolution only on the element?